### PR TITLE
[M-FSDP] Fix double buffering not working with activation recompute

### PR DESCRIPTION
**Pull request by shjwudp**
# What does this PR do ?

During activation recompute, the forward pass is interleaved within the backward pass. Previously, to prevent forward unsharded parameters from being reused by the backward function, parameter release was skipped during the re-forward phase ([megatron_fsdp.py#L781](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py#L781)). However, this caused the memory allocated by the double-buffer allocator to not be properly reclaimed in time.

This PR fixes the problem by adopting a lazy parameter release on re-forward function.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 1]
```diff
@@ -495,10 +495,10 @@ def _register_fsdp_hooks(self, root_module):
         """
         fsdp_unit_modules = self.fsdp_unit_modules
 
-        def release_module_parameters(module, *unused):
+        def release_module_parameters(module, lazy=False, *unused):
```
Comment (youngeunkwon0405) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Could you please add a comment on when `lazy` is used here?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 2]
```diff
@@ -3344,11 +3344,16 @@ def wait_bucket_ready(self, bucket_id, empty_ok=False):
         mark_bucket_ready_to_use()
 
     @torch.no_grad()
-    def release_bucket(self, bucket_id: int):
+    def release_bucket(self, bucket_id: int, lazy: bool = False):
```
Comment (youngeunkwon0405) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Can you also add a comment about the `lazy` argument? When is it used and how does it work?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: ➕

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: The issue here is that what we need is a lazy release, not no release.

Thanks for the suggestion — I’ll add some comments.

Comment (youngeunkwon0405) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: I think the comment is not that informative. 

Maybe

`lazy: If True, will release parameter buffer at [Specific release timing (TBH, I currently don't know)], otherwise it is released [Specific release timing]. This argument is used when activation recomputation is used.`

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 2 above)
Comment (youngeunkwon0405) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Here as well. I currently don't have a clear idea of when and how the bucket is released. Could you please provide more specific information?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: We can always use lazy release; the all-gather pipeline guarantees memory release at the correct time. To be conservative, we won't modify the default behavior. Therefore, `default lazy` is `False`.

I think the question is why, in some cases, the post-forward operation doesn't release model parameters. Do you think the explanation in the description is sufficient?

Comment (youngeunkwon0405) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Having a default to False is totally fine. The point is that the comment is not so informative. 

`lazy: If True, release the parameters lazily.` This provides zero additional information. Could you please provide more informative comments about this argument? 

In the above comment, I was trying to provide an example form of the comment. I guess my intention wasn't delivered well to you.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: > Could you please add a comment on when lazy is used here?

Going back to the beginning of the question, I think I've explained it clearly: we can use lazy at any time.

There is a question about why there is a mandatory lazy release in the post-forward function (when activation recompute used). The reason is that parameters can be reused during the subsequent backward function, so it is preferable to delay releasing them. I've updated the comments for post-backward; does that answer your question?

Comment (youngeunkwon0405) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: I see. Thanks for the explanation! I think my question was also not that clear, sorry about that.
I had two questions.

1. When should the `lazy` be used?
--> I think you've answered this clearly.
2. If `lazy` is used, **when** is the buffer released?
--> I can understand that it might be delayed compared to `lazy=False`, but how much will it be delayed? In this case, specifically, when will it be released? This part is not easy to infer from the code, since it is just marked as `self.bucket_can_be_released[bucket_id] = True` and handled from a different part of the code. If this part is hard to explain, I think you can leave it as is.

**Review by youngeunkwon0405 — APPROVED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: > when is the buffer released?

Buffer release occurs before all-gather pipeline requests a new buffer, and is implemented by calling `recycle_unused_buckets`.

Comment (youngeunkwon0405) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Thank you for the explanation. It will be appreciated if we can have that information somewhere near this comment.

**Review by jaredcasper — APPROVED** [review]

**Comment by cspades** [comment]
/ok to test [c30a332](https://github.com/NVIDIA/Megatron-LM/pull/2689/commits/c30a33200153c9041d6fa05d4b01faf9462f53dc)

**Comment by cspades** [comment]
So TL;DR, we're basically just trying to free some buckets in Buffer 1 at Layer N earlier (than the [post-backward](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py#L579) -> [`release_bucket`](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py#L513)) so that we can avoid synchronization delays when we allocate from Buffer 1 again at Layer N +/- 2?
